onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -441,39 +443,103 @@ static float GetQuantizationScale(const ModelBuilder& model_builder, const Node&
   return GetTensorFloatData(scale_tensor)[0];
 }
 
-static int32_t GetQuantizationZeroPoint(const ModelBuilder& model_builder, const Node& node, size_t idx) {
+static Status GetQuantizationZeroPoint(const ModelBuilder& model_builder, const Node& node, size_t idx, int32_t& zero_point) {
   std::unique_ptr<uint8_t[]> unpacked_tensor;
   size_t tensor_byte_size;
   const auto& zero_point_tensor = model_builder.GetInitializerTensors().at(node.InputDefs()[idx]->Name());
-  ORT_THROW_IF_ERROR(
+  ORT_RETURN_IF_ERROR(
       UnpackInitializerTensor(zero_point_tensor, unpacked_tensor, tensor_byte_size));
-  return static_cast<int32_t>(unpacked_tensor.get()[0]);
+  zero_point = static_cast<int32_t>(unpacked_tensor.get()[0]);
+  return Status::OK();
 }
 
-static void VerifyValidInputQuantizedType(const std::string& input_name,
-                                          const OperandType& input_operand_type,
-                                          float scale, int32_t zero_point) {
-  ORT_ENFORCE(input_operand_type.operandType.scale == scale,
-              "Input [" + input_name + "] NNAPI input: " + " scale: " +
-                  std::to_string(input_operand_type.operandType.scale) +
-                  ", ONNX input scale: " + std::to_string(scale));
+// Get scales and zero points for the qlinear binary ops (which has 2 input and 1 output)
+// QLinearConv, QLinearMatmul, QLinearAdd
+// a, b are inputs, and y is output
+static Status GetBinaryOpQuantizationScaleAndZeroPoint(const ModelBuilder& model_builder, const Node& node,
+                                                       float& a_scale, float& b_scale, float& y_scale,
+                                                       int32_t& a_zero_point, int32_t& b_zero_point, int32_t& y_zero_point) {
+  a_scale = GetQuantizationScale(model_builder, node, 1);
+  b_scale = GetQuantizationScale(model_builder, node, 4);
+  y_scale = GetQuantizationScale(model_builder, node, 6);
 
-  ORT_ENFORCE(input_operand_type.operandType.zeroPoint == zero_point,
-              "Input [" + input_name + "] NNNAPI input zero point: " +
-                  std::to_string(input_operand_type.operandType.zeroPoint) +
-                  ", ONNX input zero point: " + std::to_string(zero_point));
+  ORT_RETURN_IF_ERROR(GetQuantizationZeroPoint(model_builder, node, 2, a_zero_point));
+  ORT_RETURN_IF_ERROR(GetQuantizationZeroPoint(model_builder, node, 5, b_zero_point));
+  ORT_RETURN_IF_ERROR(GetQuantizationZeroPoint(model_builder, node, 7, y_zero_point));
+
+  return Status::OK();
+}
+
+static Status IsValidInputQuantizedType(const ModelBuilder& model_builder,
+                                        const std::string& input_name,
+                                        float scale,
+                                        int32_t zero_point) {
+  const OperandType& input_operand_type = model_builder.GetOperandTypes().at(input_name);
```

Comment:
A comment explaining what's valid would help given this seems to fail hard (throws) if not 'valid' which is slightly unexpected (naively I would have thought it the type wasn't valid it was ignored).

Reply:
Added comments, also removed the type check for TENSOR_QUANT8_ASYMM since it seems redundant

---
In reply to: [462060533](https://github.com/microsoft/onnxruntime/pull/4607#discussion_r462060533) [](ancestors = 462060533)